### ISSUE TITLE
Feedback on migrating to IAN

### ISSUE THREAD

**Issue by abrindam** [rjwut/ian#4]
We just finished migrating our custom Engineering client (https://github.com/abrindam/ArtemisEngineeringClient) from ArtClientLib to IAN and I wanted to leave some feedback.

First, the process was surprisingly painless. There were _very_ few issues, so kudos! 

Thought you might like to hear "from the field" what we went through, in case you plan to write some kind of migration document.
1. Since there is no release yet, cloned the master branch and used Eclipse to build a JAR file. Added that to our classpath and removed ArtClientLib.
2. Did a global find and replace on package names (replace "net.dhleong.acl" with "com.walkertribe.ian"). Following this, everything compiled fine.
3. Ran code, told me it needed to be pointed to an Artemis folder. For the moment, I chose to package the necessary assets inside our project. Added a few lines of code to point to this new folder.
4. No more exceptions after this, but we got stuck at our loading screen. Investigating, found that player ships are now 1-indexed instead of 0-indexed. From comments at ArtemisPlayer.getShipNumber, looks like this is intentional, so we adjusted all of our logic to account for this. As a side note, noticed that SystemManager.mPlayers array is still being initialized assuming 0-indexing, which may cause a crash on games with 8 ships.

These were the only issues we had! Hope this helps.

**Comment by rjwut** [comment]
Hey, thanks! I'll make sure to take a closer look at the stuff you're talking about here and check to see if there's anything I need to change. I intend on making an actual release once I finish the changes for Artemis 2.4.

**Comment by rjwut** [comment]
Update: I'm making some changes in IAN to accommodate packaging resources inside your JAR, so now instead of specifying a directory, you'll provide an implementation of an interface called PathResolver. IAN will come with two PathResolver implementations out of the box: FilePathResolver (loading resources from disk) and ClasspathResolver (loading resources from the classpath). So now you would do something like this:

``` java
VesselData.setPathResolver(new FilePathResolver(pathToArtemisInstall));
```

I haven't gotten around to looking into the indexing stuff, yet, but I'll do it soon.

**Comment by rjwut** [comment]
Just thought I'd let you know that I just released [IAN 3.1.0](https://github.com/rjwut/ian/releases/tag/v3.1.0), and the release addresses some of the thing you mention in this issue:
- The static `VesselData` instance has been removed. You now create a `Context` object and pass it a `PathResolver` instance which tells it how to load resources. This is more powerful because you can specify either a directory on disk (using `FilePathResolver`) or packaged within your JAR (using `ClasspathResolver`), or, if neither of those meet your needs, you can implement your own `PathResolver`. All you have to be able to do is accept a path to a resource (relative to the Artemis install path) and produce an `InputStream` for it; IAN no longer has to care exactly how the resource is loaded.
- I've addressed the issue of the zero- vs. one-based ship numbering inconsistency. This actually stems from an inconsistency in the protocol itself, but IAN now presents a consistent interface (one-based) to you so that you don't have to care which the particular underlying packet is using.

**Comment by abrindam** [comment]
Thanks! Just updated to 3.1.1 and it all looks good. It was quite easy to adapt to the new Context-based strategy.

Also loving all the detail in the release notes. Keep it up!
